File: probe/netid.py
"""Egress network fingerprint.

The public IP is a strong network identifier but is personal/locating, so
per the ethics stance we never store it raw: only a truncated SHA-256.
Resolved at most once per TTL and cached, so a heavy cycle does one lookup.
"""
from __future__ import annotations

import hashlib
import time

import httpx

_TTL_S = 600.0                 # re-resolve at most every 10 minutes
_cache: dict[str, object] = {"hash": None, "at": 0.0}


def net_hash() -> str | None:
    """Return a stable, non-reversible fingerprint of the current egress
    network (truncated hash of the public IP), or None if unavailable."""
    now = time.monotonic()
    if _cache["hash"] is not None and now - float(_cache["at"]) < _TTL_S:
        return _cache["hash"]  # type: ignore[return-value]
    try:
        ip = httpx.get("https://api.ipify.org", timeout=5.0).text.strip()
        h = hashlib.sha256(ip.encode()).hexdigest()[:12]
    except Exception:
        return _cache["hash"]  # type: ignore[return-value]  # keep last good value
    _cache["hash"], _cache["at"] = h, now
    return h

File: mobile-ios/WiFiProbeKit/Sources/WiFiProbeKit/Net/NetID.swift
import Foundation
import CryptoKit

/// Egress network fingerprint, mirroring `probe/netid.py`.
///
/// The public IP identifies a network well but is locating, so per the
/// project's ethics stance it is never stored raw: only a truncated
/// SHA-256, resolved at most once per TTL.
public actor NetID {
    public static let shared = NetID()

    private var cached: String?
    private var fetchedAt: Date?
    private let ttl: TimeInterval = 600
    private let endpoint = URL(string: "https://api.ipify.org")!

    public func hash(session: URLSession = .shared) async -> String? {
        if let cached, let fetchedAt, Date().timeIntervalSince(fetchedAt) < ttl {
            return cached
        }
        do {
            var request = URLRequest(url: endpoint)
            request.timeoutInterval = 5
            let (data, _) = try await session.data(for: request)
            guard let ip = String(data: data, encoding: .utf8)?
                .trimmingCharacters(in: .whitespacesAndNewlines), !ip.isEmpty
            else { return cached }
            cached = Self.truncatedHash(of: ip)
            fetchedAt = Date()
        } catch {
            return cached  // keep the last good value, as the probe does
        }
        return cached
    }

    static func truncatedHash(of value: String) -> String {
        SHA256.hash(data: Data(value.utf8))
            .map { String(format: "%02x", $0) }
            .joined()
            .prefix(12)
            .description
    }
}
